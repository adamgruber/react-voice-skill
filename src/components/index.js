/* eslint import/prefer-default-export: 0 */
import VoiceSkill from './VoiceSkill';

export { VoiceSkill };
