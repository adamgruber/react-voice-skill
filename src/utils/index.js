import Alexa from 'alexa-sdk';
import { EventEmitter } from 'events';
import VoiceSkill from '../components/VoiceSkill';
import IntentHandler from '../components/IntentHandler';
import Response from '../components/Response';
import Speak from '../components/Speak';
import Card from '../components/Card';

let ROOT_NODE = null;
let RESPONSE_ROOT_NODE = null;

/**
 * Updates the ref to ROOT_NODE / RESPONSE_ROOT_NODE
 *
 * @param {*} rootNode root instance
 */
function getHostContext(rootNode) {
  console.log('getHostContext');
  if (rootNode instanceof EventEmitter) {
    return ROOT_NODE = rootNode;
  }

  return RESPONSE_ROOT_NODE = rootNode;
}

/**
 * Creates an element for a skill
 *
 * @param {string} type Element type
 * @param {Object} props Component props
 * @returns {Object}
 */
function createElement(type, props) {
  console.log(`Create Element: ${type}\n`);
  // console.log(props);
  // console.log(ROOT_NODE);

  const components = {
    ROOT: () => {
      const alexa = Alexa.handler(props.event, props.context);
      alexa.appId = props.appId;
      return alexa;
    },
    RESPONSE_ROOT: () => props.handlerContext,

    VOICESKILL: () => new VoiceSkill(null, { ...props, $$type: 'VoiceSkill' }),
    INTENTHANDLER: () => new IntentHandler(ROOT_NODE, { ...props, $$type: 'IntentHandler' }),

    RESPONSE: () => new Response(RESPONSE_ROOT_NODE, { ...props, $$type: 'Response' }),
    SPEAK: () => new Speak(RESPONSE_ROOT_NODE, { ...props, $$type: 'Speak' }),
    CARD: () => new Card(RESPONSE_ROOT_NODE, { ...props, $$type: 'Card' }),
  };

  const elementCreator = components[type];

  if (elementCreator) {
    return elementCreator();
  } else {
    throw new Error(`Unknown element: ${type}`);
  }
}

export {
  createElement,
  getHostContext,
};
