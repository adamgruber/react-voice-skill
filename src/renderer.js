/* eslint no-param-reassign: 0 */
import ReactFiberReconciler from 'react-reconciler';
import { createElement, getHostContext } from './utils';

const noop = () => {};

const AlexaRenderer = ReactFiberReconciler({
  appendInitialChild(parentInstance, child) {
    console.log('appendInitialChild');
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.skill = child;
    }
  },

  createInstance(type, props, internalInstanceHandle) {
    return createElement(type, props, internalInstanceHandle);
  },

  createTextInstance(text) {
    return text;
  },

  finalizeInitialChildren() {
    return false;
  },

  getPublicInstance(instance) {
    return instance;
  },

  prepareForCommit: noop,

  prepareUpdate() {
    return true;
  },

  resetAfterCommit: noop,
  resetTextContent: noop,

  getRootHostContext(rootInstance) {
    return getHostContext(rootInstance);
  },

  getChildHostContext() {
    return {};
  },

  shouldSetTextContent() {
    return false;
  },

  now: noop,

  useSyncScheduling: true,

  mutation: {
    appendChild(parentInstance, child) {
      console.log('mutation appendInitialChild');
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        parentInstance.skill = child;
      }
    },

    appendChildToContainer(parentInstance, child) {
      if (parentInstance.appendChild) {
        parentInstance.appendChild(child);
      } else {
        switch (child.type) {
          case 'VoiceSkill':
            parentInstance.skill = child;
            break;

          case 'Response':
            parentInstance.res = child;
            break;

          default:
            // ignore
        }
      }
    },

    removeChild: noop,
    removeChildFromContainer: noop,
    insertBefore: noop,
    commitUpdate: noop,
    commitMount: noop,
    commitTextUpdate: noop,
  },
});

/**
 * @param  {ReactElement} element
 * @param {object} options
 * @param {string} options.appId Alexa App ID
 * @param {object} options.event Lambda request event
 * @param {object} options.context Lambda request context
 * @return {object}
 */
const renderToAlexa = (element, { appId, event, context }) => {
  const container = createElement('ROOT', { appId, event, context });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);
  const handlers = container.skill.render();
  console.log(handlers);
  container.registerHandlers(handlers);
  return container.execute();
};

/**
 * @param  {ReactElement} element
 * @param {object} handlerContext ASK SDK Handler Context
 * @return {object}
 */
const renderAlexaResponse = (element, handlerContext) => {
  const container = createElement('RESPONSE_ROOT', { handlerContext });
  const node = AlexaRenderer.createContainer(container);
  AlexaRenderer.updateContainer(element, node, null);

  // Get our response object from rendering components
  const response = container.res.render();
  console.log(response);

  // Parse the response object to build our Alexa response
  Object.keys(response).forEach((type) => {
    const res = response[type];
    switch (type) {
      case 'Card':
        container.response.cardRenderer(res.title, res.content, res.imgSrc);
        break;

      case 'Speak':
        container.response.speak(res.content);
        if (res.listen) {
          container.response.listen(res.listen);
        }
        break;

      default:
        // ignore
    }
  });
  return container.emit(':responseReady');
};

export {
  AlexaRenderer,
  renderToAlexa,
  renderAlexaResponse,
};
