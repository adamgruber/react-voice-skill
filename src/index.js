import { renderToVoiceSkill, renderToAlexa } from './renderer';

const Card = 'CARD';
const IntentHandler = 'INTENTHANDLER';
const Response = 'RESPONSE';
const Speak = 'SPEAK';
const VoiceSkill = 'VOICESKILL';

export {
    Card,
    IntentHandler,
    renderToAlexa,
    renderToVoiceSkill,
    Response,
    Speak,
    VoiceSkill,
};
