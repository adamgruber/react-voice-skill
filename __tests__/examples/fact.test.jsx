import React from 'react';
import { renderToAlexa } from '../../src/renderer';
import { IntentRequest, APP_ID } from '../../examples/fixtures';
import { MyVoiceSkill as App } from '../../examples/fact';

describe('HelloWorld', () => {
  let context;

  const renderApp = evt => renderToAlexa(<App />, { appId: APP_ID, event: evt, context });
  const makeEvent = intent => IntentRequest({ intent: { name: intent } });

  beforeEach(() => {
    context = {
      succeed: jest.fn(),
    };
  });

  [
    'AMAZON.HelpIntent',
    'AMAZON.CancelIntent',
    'AMAZON.StopIntent',
  ].forEach((intent) => {
    it(`should generate output for ${intent} intent`, () => {
      renderApp(makeEvent(intent));
      expect(context.succeed.mock.calls[0]).toMatchSnapshot();
    });
  });

  [
    'LaunchRequest',
    'GetNewFactIntent',
  ].forEach((intent) => {
    it(`should generate output for ${intent} intent`, () => {
      renderApp(makeEvent(intent));
      expect(context.succeed.mock.calls[0]).toEqual([expect.objectContaining({
        response: {
          card: {
            content: expect.any(String),
            title: 'Space Facts',
            type: 'Simple',
          },
          outputSpeech: {
            ssml: expect.any(String),
            type: 'SSML',
          },
          reprompt: {
            outputSpeech: {
              ssml: '<speak> What did you say? </speak>',
              type: 'SSML',
            },
          },
          shouldEndSession: false,
        },
        sessionAttributes: {},
        userAgent: 'ask-nodejs/1.0.25 Node/v8.9.0',
        version: '1.0',
      })]);
    });
  });
});
