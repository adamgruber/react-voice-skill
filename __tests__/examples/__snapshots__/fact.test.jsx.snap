// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWorld should generate output for AMAZON.CancelIntent intent 1`] = `
Array [
  Object {
    "response": Object {
      "outputSpeech": Object {
        "ssml": "<speak> Goodbye! </speak>",
        "type": "SSML",
      },
      "shouldEndSession": true,
    },
    "sessionAttributes": Object {},
    "userAgent": "ask-nodejs/1.0.25 Node/v8.9.0",
    "version": "1.0",
  },
]
`;

exports[`HelloWorld should generate output for AMAZON.HelpIntent intent 1`] = `
Array [
  Object {
    "response": Object {
      "outputSpeech": Object {
        "ssml": "<speak> You can say tell me a space fact, or, you can say exit... What can I help you with? </speak>",
        "type": "SSML",
      },
      "reprompt": Object {
        "outputSpeech": Object {
          "ssml": "<speak> What can I help you with? </speak>",
          "type": "SSML",
        },
      },
      "shouldEndSession": false,
    },
    "sessionAttributes": Object {},
    "userAgent": "ask-nodejs/1.0.25 Node/v8.9.0",
    "version": "1.0",
  },
]
`;

exports[`HelloWorld should generate output for AMAZON.StopIntent intent 1`] = `
Array [
  Object {
    "response": Object {
      "outputSpeech": Object {
        "ssml": "<speak> Goodbye! </speak>",
        "type": "SSML",
      },
      "shouldEndSession": true,
    },
    "sessionAttributes": Object {},
    "userAgent": "ask-nodejs/1.0.25 Node/v8.9.0",
    "version": "1.0",
  },
]
`;
